Regression in version 2.20.x

Calling ` mvn versions:display-dependency-updates -Dmaven.version.ignore="Alpha"`

Generates following exception after upgrading from 2.19.1 to 2.20.0 (and 2.20.1).

```
Execution default-cli of goal org.codehaus.mojo:versions-maven-plugin:2.20.0:display-dependency-updates failed: Cannot invoke "java.lang.CharSequence.length()" because "this.text" is null
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2 (MojoExecutor.java:333)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:316)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:212)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:174)
    at org.apache.maven.lifecycle.internal.MojoExecutor.access$000 (MojoExecutor.java:75)
    at org.apache.maven.lifecycle.internal.MojoExecutor$1.run (MojoExecutor.java:162)
    at org.apache.maven.plugin.DefaultMojosExecutionStrategy.execute (DefaultMojosExecutionStrategy.java:39)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:159)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:105)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:73)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:53)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:118)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:261)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:173)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:101)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:906)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:283)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:206)
    at jdk.internal.reflect.DirectMethodHandleAccessor.invoke (DirectMethodHandleAccessor.java:103)
    at java.lang.reflect.Method.invoke (Method.java:580)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:255)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:201)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:361)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:314)
Caused by: org.apache.maven.plugin.PluginExecutionException: Execution default-cli of goal org.codehaus.mojo:versions-maven-plugin:2.20.0:display-dependency-updates failed: Cannot invoke "java.lang.CharSequence.length()" because "this.text" is null
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:133)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2 (MojoExecutor.java:328)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:316)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:212)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:174)
    at org.apache.maven.lifecycle.internal.MojoExecutor.access$000 (MojoExecutor.java:75)
    at org.apache.maven.lifecycle.internal.MojoExecutor$1.run (MojoExecutor.java:162)
    at org.apache.maven.plugin.DefaultMojosExecutionStrategy.execute (DefaultMojosExecutionStrategy.java:39)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:159)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:105)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:73)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:53)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:118)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:261)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:173)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:101)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:906)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:283)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:206)
    at jdk.internal.reflect.DirectMethodHandleAccessor.invoke (DirectMethodHandleAccessor.java:103)
    at java.lang.reflect.Method.invoke (Method.java:580)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:255)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:201)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:361)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:314)
Caused by: java.lang.NullPointerException: Cannot invoke "java.lang.CharSequence.length()" because "this.text" is null
    at java.util.regex.Matcher.getTextLength (Matcher.java:1808)
    at java.util.regex.Matcher.reset (Matcher.java:461)
    at java.util.regex.Matcher.<init> (Matcher.java:256)
    at java.util.regex.Pattern.matcher (Pattern.java:1180)
    at org.codehaus.mojo.versions.api.IgnoreVersionHelper.isVersionIgnoredRegex (IgnoreVersionHelper.java:123)
    at org.codehaus.mojo.versions.api.IgnoreVersionHelper.access$100 (IgnoreVersionHelper.java:35)
    at org.codehaus.mojo.versions.api.IgnoreVersionHelper$IgnoreVersionType.lambda$TYPE_REGEX$0 (IgnoreVersionHelper.java:52)
    at org.codehaus.mojo.versions.api.IgnoreVersionHelper.lambda$isVersionIgnored$0 (IgnoreVersionHelper.java:103)
    at java.util.Optional.map (Optional.java:260)
    at org.codehaus.mojo.versions.api.IgnoreVersionHelper.isVersionIgnored (IgnoreVersionHelper.java:103)
    at org.codehaus.mojo.versions.rule.RuleServiceUtils.lambda$filterByRuleService$1 (RuleServiceUtils.java:39)
    at java.util.stream.MatchOps$1MatchSink.accept (MatchOps.java:90)
    at java.util.ArrayList$ArrayListSpliterator.tryAdvance (ArrayList.java:1685)
    at java.util.stream.ReferencePipeline.forEachWithCancel (ReferencePipeline.java:129)
    at java.util.stream.AbstractPipeline.copyIntoWithCancel (AbstractPipeline.java:527)
    at java.util.stream.AbstractPipeline.copyInto (AbstractPipeline.java:513)
    at java.util.stream.AbstractPipeline.wrapAndCopyInto (AbstractPipeline.java:499)
    at java.util.stream.MatchOps$MatchOp.evaluateSequential (MatchOps.java:230)
    at java.util.stream.MatchOps$MatchOp.evaluateSequential (MatchOps.java:196)
    at java.util.stream.AbstractPipeline.evaluate (AbstractPipeline.java:234)
    at java.util.stream.ReferencePipeline.noneMatch (ReferencePipeline.java:642)
    at org.codehaus.mojo.versions.rule.RuleServiceUtils.lambda$filterByRuleService$0 (RuleServiceUtils.java:38)
    at org.codehaus.mojo.versions.api.ArtifactVersions.filter (ArtifactVersions.java:91)
    at org.codehaus.mojo.versions.rule.RuleServiceUtils.filterByRuleService (RuleServiceUtils.java:38)
    at org.codehaus.mojo.versions.internal.DependencyUpdatesLoggingHelper.getDependencyUpdates (DependencyUpdatesLoggingHelper.java:107)
    at org.codehaus.mojo.versions.DisplayDependencyUpdatesMojo.logUpdates (DisplayDependencyUpdatesMojo.java:555)
    at org.codehaus.mojo.versions.DisplayDependencyUpdatesMojo.execute (DisplayDependencyUpdatesMojo.java:478)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:126)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2 (MojoExecutor.java:328)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:316)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:212)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:174)
    at org.apache.maven.lifecycle.internal.MojoExecutor.access$000 (MojoExecutor.java:75)
    at org.apache.maven.lifecycle.internal.MojoExecutor$1.run (MojoExecutor.java:162)
    at org.apache.maven.plugin.DefaultMojosExecutionStrategy.execute (DefaultMojosExecutionStrategy.java:39)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:159)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:105)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:73)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:53)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:118)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:261)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:173)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:101)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:906)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:283)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:206)
    at jdk.internal.reflect.DirectMethodHandleAccessor.invoke (DirectMethodHandleAccessor.java:103)
    at java.lang.reflect.Method.invoke (Method.java:580)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:255)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:201)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:361)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:314)
```

--- a/versions-common/src/main/java/org/codehaus/mojo/versions/api/ArtifactVersions.java
+++ b/versions-common/src/main/java/org/codehaus/mojo/versions/api/ArtifactVersions.java
@@ -88,7 +88,7 @@ public class ArtifactVersions extends AbstractVersionDetails implements Comparab
         ArtifactVersions result = new ArtifactVersions(
                 artifact,
                 versions.stream().filter(v -> versionFilter.test(v.toString())).collect(Collectors.toList()));
-        if (!versionFilter.test(result.artifact.getVersion())) {
+        if (result.artifact.getVersion() != null && !versionFilter.test(result.artifact.getVersion())) {
             result.setCurrentVersion((ArtifactVersion) null);
         }
         return result;

--- a/versions-common/src/main/java/org/codehaus/mojo/versions/api/IgnoreVersionHelper.java
+++ b/versions-common/src/main/java/org/codehaus/mojo/versions/api/IgnoreVersionHelper.java
@@ -15,6 +15,8 @@ package org.codehaus.mojo.versions.api;
  * limitations under the License.
  */
 
+import javax.annotation.Nonnull;
+
 import java.util.Arrays;
 import java.util.List;
 import java.util.Optional;
@@ -92,13 +94,17 @@ public class IgnoreVersionHelper {
     }
 
     /**
-     * Check if the provided version is ignored by the provided {@link IgnoreVersion} instance
+     * <p>Check if the provided version is ignored by the provided {@link IgnoreVersion} instance.</p>
+     * <p>In the rare case that the {@code v} should be be {@code null}, the function returns {@code true}.</p>
      *
      * @param v version string to be checked
      * @param ignoreVersion {@link IgnoreVersion} instance providing the filter
-     * @return {@code true} if the provided version is ignored
+     * @return {@code true} if the provided version is ignored or if it is equal to {@code null}.
      */
     public static boolean isVersionIgnored(String v, IgnoreVersion ignoreVersion) {
+        if (v == null) {
+            return true;
+        }
         return IgnoreVersionType.forType(ignoreVersion.getType())
                 .map(t -> t.getPredicate().apply(v, ignoreVersion))
                 .orElseThrow(() -> new IllegalArgumentException("Invalid version type: " + ignoreVersion.getType()));
@@ -111,7 +117,7 @@ public class IgnoreVersionHelper {
      * @param ignoreVersion {@link IgnoreVersion} instance providing the filter
      * @return {@code true} if the provided version is ignored
      */
-    public static boolean isVersionIgnored(Version v, IgnoreVersion ignoreVersion) {
+    public static boolean isVersionIgnored(@Nonnull Version v, IgnoreVersion ignoreVersion) {
         return isVersionIgnored(v.toString(), ignoreVersion);
     }
 
@@ -119,11 +125,11 @@ public class IgnoreVersionHelper {
         return ignoreVersion.getVersion().equals(v);
     }
 
-    private static boolean isVersionIgnoredRegex(String v, IgnoreVersion ignoreVersion) {
+    private static boolean isVersionIgnoredRegex(@Nonnull String v, IgnoreVersion ignoreVersion) {
         return Pattern.compile(ignoreVersion.getVersion()).matcher(v).matches();
     }
 
-    private static boolean isVersionIgnoredRange(String v, IgnoreVersion ignoreVersion) {
+    private static boolean isVersionIgnoredRange(@Nonnull String v, IgnoreVersion ignoreVersion) {
         try {
             ArtifactVersion aVersion = ArtifactVersionService.getArtifactVersion(v);
             VersionRange versionRange = VersionRange.createFromVersionSpec(ignoreVersion.getVersion());

--- a/versions-common/src/test/java/org/codehaus/mojo/versions/api/IgnoreVersionHelperTest.java
+++ b/versions-common/src/test/java/org/codehaus/mojo/versions/api/IgnoreVersionHelperTest.java
@@ -27,6 +27,7 @@ import org.junit.jupiter.params.provider.MethodSource;
 
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
 import static org.junit.jupiter.params.provider.Arguments.of;
 
 class IgnoreVersionHelperTest {
@@ -62,6 +63,15 @@ class IgnoreVersionHelperTest {
 
         assertEquals(
                 expected, IgnoreVersionHelper.isVersionIgnored(VERSION_SCHEME.parseVersion(version), ignoreVersion));
+    }
+
+    /*
+     * Additional safety measure: prevent a NPE coming from Matcher if the current version happens to be null.
+     */
+    @Test
+    public void testNullIgnoreVersion() {
+        IgnoreVersion ignoreVersion = aIgnoreVersion("Alpha", IgnoreVersion.TYPE_REGEX);
+        assertTrue(IgnoreVersionHelper.isVersionIgnored((String) null, ignoreVersion));
     }
 
     public static Stream<Arguments> ignoredTypeRange() {

--- a/versions-maven-plugin/src/test/java/org/codehaus/mojo/versions/DisplayDependencyUpdatesMojoTest.java
+++ b/versions-maven-plugin/src/test/java/org/codehaus/mojo/versions/DisplayDependencyUpdatesMojoTest.java
@@ -19,6 +19,7 @@ import java.io.File;
 import java.nio.charset.Charset;
 import java.nio.file.Files;
 import java.util.Arrays;
+import java.util.Collections;
 import java.util.HashMap;
 import java.util.List;
 
@@ -731,6 +732,33 @@ public class DisplayDependencyUpdatesMojoTest extends AbstractMojoTestCase {
             }.execute();
 
             assertThat(String.join("", Files.readAllLines(tempFile.getPath())), not(containsString("1.0.0")));
+        }
+    }
+
+    /*
+     * Regression guard: in case a dependency artifact has no "immediate" version defined, which can happen if the version
+     * is defined as a restriction, ArtifactVersions::filter would still call the IgnoreVersionHelper on the current
+     * version, which is null, causing a NullPointerException.
+     */
+    @Test
+    public void testIgnoredVersionsWithNoCurrentVersion() throws Exception {
+        try (CloseableTempFile tempFile = new CloseableTempFile("display-dependency-updates")) {
+
+            DisplayDependencyUpdatesMojo mojo = mojoRule.lookupConfiguredMojo(
+                    new File("target/test-classes/org/codehaus/mojo/display-dependency-updates/ranges"),
+                    "display-dependency-updates");
+            mojo.ignoredVersions = Collections.singleton("Alpha");
+            mojo.outputFile = tempFile.getPath().toFile();
+            mojo.setPluginContext(new HashMap<>());
+            mojo.repositorySystem = mockAetherRepositorySystem(new HashMap<String, String[]>() {
+                {
+                    put("dummy-api", new String[] {"1.0.0"});
+                }
+            });
+
+            mojo.execute();
+            List<String> output = Files.readAllLines(tempFile.getPath(), UTF_8);
+            assertThat(output, not(hasItem(containsString("1.0.0"))));
         }
     }
 }

--- a/versions-maven-plugin/src/test/resources/org/codehaus/mojo/display-dependency-updates/ranges/pom.xml
+++ b/versions-maven-plugin/src/test/resources/org/codehaus/mojo/display-dependency-updates/ranges/pom.xml
@@ -1,0 +1,17 @@
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+  <groupId>default-group</groupId>
+  <artifactId>default-artifact</artifactId>
+  <version>1.0</version>
+  <packaging>pom</packaging>
+
+  <dependencies>
+    <dependency>
+      <groupId>localhost</groupId>
+      <artifactId>dummy-api</artifactId>
+      <version>(,2.0.0)</version>
+    </dependency>
+  </dependencies>
+
+</project>

